packages/aws-cdk-lib/aws-cloudfront/lib/function.ts

```diff
@@ -196,11 +197,17 @@ export class Function extends Resource implements IFunction {
   }
 
   private generateName(): string {
-    const name = Stack.of(this).region + Names.uniqueId(this);
-    if (name.length > 64) {
-      return name.substring(0, 32) + name.substring(name.length - 32);
-    }
-    return name;
+    // Function names are globally unique, even if the stack is deployed in different regions;
+    // therefore, we prefix the region name to the resource name. It's important; however, to make
+    // sure that we don't exceed the service's length limit. We subtract the length of the region
+    // name from the max length to support the concatenation. If the region is currently unresolved,
+    // then we use the name of the longest known region.
+    const serviceMaxNameLength = 64;
+    const region = Stack.of(this).region;
+    const regionNameLength = !Token.isUnresolved(region) ? region.length : Math.max(...Fact.regions.map((r) => r.length));
```

Comment:
It seems to me that this can potentially trigger re-deploy for those cases where `region` is not resolved.
Given that the issue needs to be fixed and it's probably going to be a breaking change anyway, I think it may be worth considering a refactor/simplification of the name generation function (under a feature flag) to avoid future issues, eg:
```ts
return Lazy.string({
  produce: () => Names.uniqueResourceName(this, { maxLength: 64, allowedSpecialCharacters: '-_' }),
});
```

Reply:
I think that including the region name is, unfortunately, a very intentional decision so that the same stack can be deployed in multiple regions. https://github.com/aws/aws-cdk/pull/14511#discussion_r625681979

But then again, any callers can always just pass ``functionName: `${this.env.region}Name`,`` if that's really an use case for them under the feature flag.

Reply:
Added the feature flag and including a note about region inclusion as part of the `detailsMd`. Let me know what you think!